### PR TITLE
Add a factory for Poisson / NegativeBinomial / Binomial / BetaBinomial

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2426 

This creates a new helper `infection_dist()` that creates one of a Poisson, NegativeBinomial, Binomial, or BetaBinomial depending on the quantitative values of its arguments. Of these four distributions, I intend only to use the Binomial and BetaBinomial; the others are present more as a bridge to limiting parameterizations that are more common in epidemiology.

image

This PR also refactors the SIR and SEIR models to use the new helper. This refactoring prepares for overdispersed versions of these distributions in a follow-up PR #2451.

## Tested
- [x] refactoring is covered by existing tests
- [x] added a new unit test confirming approximate agreement in four limiting cases
- [x] visually validated agreement in [a notebook](https://github.com/fritzo/notebooks/blob/master/infection_dist.ipynb)
- [x] ran `sir.py -p 10000 -d 60 -f 30 --plot` as a sanity check

### CODE REVIEW

File: pyro/contrib/epidemiology/distributions.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            # Return an overdispersed Negative-Binomial distribution.
+            combined_k = k * I
+            logits = torch.as_tensor(R / k).log()
+            return dist.NegativeBinomial(combined_k, logits=logits)
+    else:
+        # Compute the probability that any given (susceptible, infectious)
+        # pair of individuals results in an infection at this time step.
+        p = torch.as_tensor(R / N).clamp(max=1 - 1e-6)
+        # Combine infections from all individuals.
+        combined_p = p.neg().log1p().mul(I).expm1().neg()  # = 1 - (1 - p)**I
```
Comment (fehiepsi) [inline pyro/contrib/epidemiology/distributions.py:85]: I think this is not consistent w.r.t. before. Previously, `combined_p` is `1 - (e^-rate_s) ** I = 1 - (e^-p) ** I`.

File: pyro/contrib/epidemiology/distributions.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    distribution depending on whether ``population`` and ``concentration`` are
+    finite. In Pyro models, the population is usually finite. In the limit
+    ``population → ∞`` and ``num_susceptible/population → 1``, the Binomial
+    converges to Poisson and the Beta-Binomial converges to Negative-Binomial.
+    In the limit ``concentration → ∞``, the Negative-Binomial converges to
+    Poisson and the Beta-Binomial converges to Binomial.
+
+    The overdispersed distributions (Negative-Binomial and Beta-Binomial
+    returned when ``concentration < ∞``) are useful for modeling superspreader
+    individuals [2,3]. The finitely supported distributions Binomial and
```
Comment (martinjankowiak) [inline pyro/contrib/epidemiology/distributions.py:31]: are these the right refs?

Comment (martinjankowiak) [reply pyro/contrib/epidemiology/distributions.py:31]: => [1, 2]?

File: pyro/contrib/epidemiology/distributions.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        ``R0 / tau`` where ``R0`` is the basic reproductive number and ``tau``
+        is the mean duration of infectiousness.
+    :param num_infectious: The number of infectious individuals at this
+        time step, sometimes ``I``, sometimes ``E+I``.
+    :param num_susceptible: The number ``S`` of susceptible individuals at this
+        time step. This defaults to an infinite population.
+    :param population: The total number of individuals in a population.
+        This defaults to an infinite population.
+    :concentration: The concentration or dispersion parameter ``k`` in
+        overdispersed models of superspreaders [2,3]. This defaults to minimum
```
Comment (martinjankowiak) [inline pyro/contrib/epidemiology/distributions.py:61]: same here?

File: pyro/contrib/epidemiology/distributions.py (hunk 1 above)
Comment (fritzo) [reply pyro/contrib/epidemiology/distributions.py:85]: That's right, I am changing the formula. This PR seems to make it more mathematically plausible.

Comment (fritzo) [reply pyro/contrib/epidemiology/distributions.py:85]: @fehiepsi I am not super confident in these formulas, but the plots do show they are at least in agreement. Let me know if you have any suggestions.

Comment (fehiepsi) [reply pyro/contrib/epidemiology/distributions.py:85]: This seems to be similar to [Reed-Frost model](https://en.wikipedia.org/wiki/Reed%E2%80%93Frost_model#Mathematics).

**Comment by fritzo** [comment]
Thanks for reviewing!

**Pull request merged**